### ISSUE TITLE
Datastore save() fails on null value when excludeFromIndexes is used for nested arrays

### ISSUE THREAD

**Issue by gunsch** [googleapis/nodejs-datastore#93]
#### Environment details

  - OS: Debian (rodete)
  - Node.js version: 8.9.4
  - npm version: 5.6.0
  - `@google-cloud/datastore` version: 1.4.0

#### Steps to reproduce

Execute the following code (snippet taken from a tiny express app to demo the issue):

```javascript
router.get('/save', function(req, res, next)  {
  const datastoreClient = Datastore();
  const entity = {
    key: datastoreClient.key(['test']),
    data: {
      "rules": [
        {"requirements": [{"audiences": "value"}]},
        {"requirements": null},
      ]
    },
    excludeFromIndexes: [
      'rules[].requirements[].audiences'
    ]
  };
  datastoreClient.save(entity)
      .then(() => { res.send('OK!'); })
      .catch(error => { res.send('Failed: ' + error); });
});
```

Result:
`Failed: TypeError: Cannot read property 'values' of undefined`

With some digging, it seems it's failing here:
https://github.com/googleapis/nodejs-datastore/blob/c1cfe5bfce2ba018c3d2b4e5d03fd1c6e2d420b8/src/entity.js#L539

This only happens with the combination of this data structure + this `excludeFromIndexes` value, since it causes a traversal of the object, including an expectation that the second `requirements` field is also an array.

It's sort of a weird edge case, but one we encountered in production this week.

**Comment by beaulac** [comment]
This is a known limitation of the current `excludeFromIndexes` syntax. https://github.com/GoogleCloudPlatform/google-cloud-node/pull/2773

**Comment by trollkotze** [comment]
I have included a fix for this in PR #101, which was originally made to fix #100.
This now allows things like this:
```
entity = {
  key: key,
  excludeFromIndexes: ['arr[].a[].b'],
  data: {
    nestedArrayVariants: [
      {
        a: [{b:'something'},{c:'else'}],
      },{
        a: null,
      },{
        a: 'a string',
      },{
        a: ['whatever'],
      }
    ],
  },
};
```
Instead of throwing a type error, it is now simply recognized that the exclusion rule does not apply for the elements of `arr` where `arr[i].a` is not an array.

Similarly for the root,
```
 excludeFromIndexes: ['a[]']
```
will now only take effect in array fashion (i.e. exclude all direct elements of `a` from indexes) if the property 'a' is actually an array value.

On the other hand,
```
  excludeFromIndexes: ['a']
```
does still not check whether `a` is *not* an array. If it *is* an array value, then `excludeFromIndexes = true` will be set at the root of `a` in the entityProto, instead of for each individual element of that array property. This is not allowed and results in an `Error: 3 INVALID_ARGUMENT: Exclude from indexes cannot be set on a list value` being thrown on the network request.

I think for consistency, as well as flexibility, it would be best, in case of exclusion paths ending in non-array properties, like
```
excludeFromIndexes: 'path.to[].some.property'
```
, to apply them only where the end node `<...>.property` is actually not an array.
For array properties, we have `excludeFromIndexes: ['<...>.property[]']`, which is simply not applied if `<...>.property` is *not* an array.
Likewise, `excludeFromIndexes: ['<...>.property']` should simply not be applied where `<...>.property` is an array.

**Comment by trollkotze** [comment]
The problem described in previous message is now handled in the latest commit to PR #101:
[Commit](https://github.com/googleapis/nodejs-datastore/pull/101/commits/756711743f20f1e2329fedd23b09549249d8a78e)
